File: exercicios/projeto-final/skeletons/pokeapi/react-native/pratica/src/types/pokemon.ts
export interface PokemonSummary {
  id: number;
  name: string;
}

export interface PokemonDetail {
  id: number;
  name: string;
  types: string[];
  heightDm: number;
  weightHg: number;
}

export function spriteUrl(id: number): string {
  return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${id}.png`;
}

export function idFromUrl(url: string): number {
  const segments = url.split('/').filter(Boolean);
  return Number(segments[segments.length - 1]);
}

export function capitalize(value: string): string {
  return value.length === 0 ? value : value[0].toUpperCase() + value.slice(1);
}

File: exercicios/projeto-final/skeletons/pokeapi/flutter/pratica/lib/models/pokemon.dart
class PokemonSummary {
  final int id;
  final String name;

  PokemonSummary({required this.id, required this.name});

  String get spriteUrl =>
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/$id.png';

  factory PokemonSummary.fromListEntry(String name, String url) {
    final segments = url.split('/').where((s) => s.isNotEmpty).toList();
    final id = int.parse(segments.last);
    return PokemonSummary(id: id, name: name);
  }
}

class PokemonDetail {
  final int id;
  final String name;
  final List<String> types;
  final int heightDm;
  final int weightHg;

  PokemonDetail({
    required this.id,
    required this.name,
    required this.types,
    required this.heightDm,
    required this.weightHg,
  });

  String get spriteUrl =>
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/$id.png';

  factory PokemonDetail.fromJson(Map<String, dynamic> json) {
    final types = (json['types'] as List)
        .map((t) => t['type']['name'] as String)
        .toList();
    return PokemonDetail(
      id: json['id'] as int,
      name: json['name'] as String,
      types: types,
      heightDm: json['height'] as int,
      weightHg: json['weight'] as int,
    );
  }
}
